### PR TITLE
perf(backend): O(1) ALLOWED_DTYPES membership and a fast path for canonical dtype strings

### LINKED ISSUE #23272: standardize_dtype / result_type do uncached O(n) dtype lookups
`standardize_dtype` (`keras/src/backend/common/variables.py`) and `result_type` (`keras/src/backend/common/dtypes.py`) sit on the per-op dtype-resolution hot path and are called on essentially every tensor op — in the benchmark used for this series, `standardize_dtype` alone fires roughly 177 times per single LLM forward pass.

The work is redundant per call because the input space is small and mostly fixed: `standardize_dtype` re-ran a full resolution pipeline (including a linear membership scan over the `ALLOWED_DTYPES` tuple) even when it was handed an already-canonical dtype string like `"float32"`, and `result_type` re-derived its result from scratch on every call even though its inputs are drawn from the same small, fixed set of dtypes for the lifetime of a model. Neither result changes between calls with the same arguments, so recomputing is wasted work.

Fixed by #23182: `ALLOWED_DTYPES` membership is now checked via an O(1) `frozenset` instead of a linear tuple scan, and both `standardize_dtype` and `result_type` gained bounded/`lru_cache`-based caches so repeat calls with the same input skip the resolution pipeline entirely.

Part of the per-call Python-dispatch overhead series in #22561. Fixed by #23182.

### PR DESCRIPTION
**Pull request by pctablet505**
Closes #23272

## Summary

`standardize_dtype` runs a chain of `hasattr` probes and a linear membership test against the `ALLOWED_DTYPES` tuple on every call, even when it is handed a dtype string that is already canonical — which is the overwhelmingly common case.

Two changes:

- `dtypes.ALLOWED_DTYPES_SET`, a `frozenset` derived from `ALLOWED_DTYPES`, so the membership test is O(1). `ALLOWED_DTYPES` itself stays an ordered tuple: several test suites build parametrized cases by iterating it, and pytest-xdist needs a deterministic collection order across workers.
- An early return for a `str` that is already in that set, skipping the normalization chain entirely.

Everything else is unchanged.

## Effect

Single process, interleaved with an A/A control, 20k iterations × 11 rounds:

| input | master | this PR |
|---|--:|--:|
| `"float32"` (canonical string) | 110.0 ns | **32.3 ns** |
| A/A control | 109.7 ns | 32.2 ns |
| `torch.float32` (non-string path) | 208.0 ns | 214.4 ns |

3.4× on the common path. The non-string path pays one extra `isinstance` check and is ~3% slower; the A/A control puts the noise floor under 1%, so that cost is real but small.

End to end, `standardize_dtype` is called 71 times per forward pass on a small CNN and 297 times on a 20-layer MLP (counted with `cProfile`, which is immune to `from … import` rebinding). Measured PR vs master, CPU, medians of interleaved runs:

| model | master | this PR | |
|---|--:|--:|---|
| MLP, 20 layers | 1.3170 ms | 1.2992 ms | **−1.4%** (master spread 0.4%) |
| CNN, 2 blocks | 3.2308 ms | 3.1950 ms | −1.1% (master spread 5.0%, so within noise) |

The MLP figure is outside its noise band and agrees with the arithmetic (297 × 77.7 ns = 23 µs on a 1.32 ms step). The CNN does proportionally more real compute per layer call, so the effect disappears into noise there.

## Equivalence

A differential probe of 48 inputs — every `ALLOWED_DTYPES` string, every `PYTHON_DTYPES_MAP` key, `None`, Python scalar types, numpy dtypes and scalar types, torch dtypes, `"torch.*"` / `"jax.numpy.*"` strings, `"torch.float32.extra"`, and invalid inputs including `1`, `True`, `[]`, `{}` and `object()` — returns identical results and identical exception types and messages on both trees. The only differing byte is the memory address inside the `object()` error message, which differs between any two processes.

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

**Comment by hertschuh** [comment]
Is this still in draft? Or is this ready for review?

**Comment by pctablet505** [comment]
> Is this still in draft? Or is this ready for review?

Hi, It is ready for review now.

File: keras/src/backend/common/variables.py [hunk 1]
```diff
@@ -567,25 +567,69 @@ def initialize_all_variables():
...
+        # A mapped alias like "int" falls through to PYTHON_DTYPES_MAP.
+        mapped = dtypes.PYTHON_DTYPES_MAP.get(dtype)
+        if mapped is not None:
+            return mapped
+        # Framework-prefixed strings like "torch.float32" or
+        # "jax.numpy.float32" resolve by splitting on ".".
+        if "torch" in dtype or "jax.numpy" in dtype:
+            resolved = dtype.split(".")[-1]
+            if resolved in dtypes.ALLOWED_DTYPES_SET:
+                return resolved
```
Comment (hertschuh) [inline keras/src/backend/common/variables.py:592]: This was not done this way before.

Before this was done **after** calling `__str__` on the object.

File: keras/src/backend/common/variables.py [hunk 2]
```diff
@@ -567,25 +567,69 @@ def initialize_all_variables():
...
+    # Fast path 2: numpy dtype object. .name is always canonical; uint32
+    # is NOT remapped to int64.
+    if isinstance(dtype, np.dtype):
+        name = dtype.name
+        if name in dtypes.ALLOWED_DTYPES_SET:
+            return name
+        raise ValueError(f"Invalid dtype: {name}")
+
+    # Fast path 3: cached non-string objects (torch dtypes, Python types).
+    cache_key = (type(dtype), dtype)
```
Comment (hertschuh) [inline keras/src/backend/common/variables.py]: So this weird `(type(dtype), dtype)` is just for TensorFlow?

Can you cache the dtypes for the other backends normally?

File: keras/src/backend/common/dtypes.py [hunk 3]
```diff
@@ -277,6 +283,26 @@ def _lattice_result_type(*args):
...
+        return floatx
+    for dtype in dtypes:
+        if dtype in FLOAT8_TYPES:
+            raise ValueError(
+                "There is no implicit conversions from float8 dtypes to others."
+                f" You must cast it internally. Received: {dtypes}"
+            )
+    return _lattice_result_type(
+        *(floatx if arg is None else arg for arg in dtypes),
+    )
```
Comment (hertschuh) [inline keras/src/backend/common/dtypes.py]: Can you carve out this part (optimizing `result_type` and related tests) in a separate PR?

I'd like to address `ALLOWED_DTYPES` and `standardize_dtype` first.

Comment (pctablet505) [reply keras/src/backend/common/dtypes.py]: https://github.com/keras-team/keras/pull/23182 
Created a new PR for it.

File: keras/src/backend/common/variables.py (hunk 1 above)
Comment (pctablet505) [reply keras/src/backend/common/variables.py:592]: It did move, but for a `str` input it's the same test. `str(dtype) is dtype` for a
string, and a `str` has neither `.name` nor `.__name__`, so on master a string always
fell through the first two branches and landed on exactly this `__str__` check — it was
the only branch a string could ever reach. Reordering it can't change the result.

The other reorder is safe for the same kind of reason: the only string key in
`PYTHON_DTYPES_MAP` is `"int"`, which isn't in `ALLOWED_DTYPES`, so no string can match
both lookups.

Error messages are preserved too — `"torch.float32.extra"` still raises
`Invalid dtype: extra`, same as master, and that case is in the tests along with
`"torch.float32"` and `torch.float32`.

Non-string objects are untouched; they still go through the original `__str__` chain
further down.

File: keras/src/backend/common/variables.py (hunk 2 above)
Comment (pctablet505) [reply keras/src/backend/common/variables.py]: Yes, TensorFlow is the only reason. `tf.float32 == 1` and `hash(tf.float32) == hash(1)`,
so with a bare key, caching `tf.float32 -> "float32"` makes a later `standardize_dtype(1)`
return `"float32"` instead of raising. `True` collides the same way.

Splitting it into a TF cache and a normal one needs a per-call check for which one to
use, and that's the part I'd rather avoid: `tf` in this module is the lazy proxy from
`keras.src.utils.module_utils`, so `isinstance(dtype, tf.DType)` would import and
initialize TensorFlow on the torch/jax/numpy backends just to pick a dict. The tf-free
alternative, testing `type(dtype).__module__`, costs about what building the tuple costs.

So the tuple is one small allocation, no backend branch, and correct everywhere. Happy to
switch to the explicit split if you'd prefer it — just wanted to explain why it looks the
way it does.

File: keras/src/backend/common/variables.py [hunk 4]
```diff
@@ -567,25 +567,62 @@ def initialize_all_variables():
...
+_DTYPE_CACHE = {}
+
+
 @keras_export(
     ["keras.utils.standardize_dtype", "keras.backend.standardize_dtype"]
 )
 def standardize_dtype(dtype):
+    if isinstance(dtype, str):
+        if dtype in dtypes.ALLOWED_DTYPES_SET:
+            return dtype
```
Comment (hertschuh) [inline keras/src/backend/common/variables.py:582]: How much speed-up to you get if you undo everything, but keep `ALLOWED_DTYPES_SET`, add these 3 lines of code...

File: keras/src/backend/common/variables.py [hunk 5]
```diff
@@ -567,25 +567,62 @@ def initialize_all_variables():
...
+        "torch" in str(resolved) or "jax.numpy" in str(resolved)
     ):
-        dtype = str(dtype).split(".")[-1]
+        resolved = str(resolved).split(".")[-1]
 
-    if dtype not in dtypes.ALLOWED_DTYPES:
-        raise ValueError(f"Invalid dtype: {dtype}")
-    return dtype
+    if resolved not in dtypes.ALLOWED_DTYPES_SET:
+        raise ValueError(f"Invalid dtype: {resolved}")
```
Comment (hertschuh) [inline keras/src/backend/common/variables.py:622]: ... and keep this change (from `ALLOWED_DTYPE` to `ALLOWED_DTYPES_SET`)?

**Review by hertschuh — APPROVED** [review]
Thanks!
